internal/cache/redis.go

```diff
@@ -0,0 +1,174 @@
+// Package cache provides application and HTTP response cache.
+package cache
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/RHEnVision/provisioning-backend/internal/clients"
+	"github.com/RHEnVision/provisioning-backend/internal/config"
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/metrics"
+	"github.com/RHEnVision/provisioning-backend/internal/models"
+	"github.com/redis/go-redis/v9"
+	"github.com/rs/zerolog/log"
+)
+
+var (
+	ErrNotFound    = errors.New("not found in cache")
+	ErrUnknownType = errors.New("cannot find cache prefix")
+
+	// when redis is enabled via configuration
+	redisEnabled bool
+
+	// the client
+	client *redis.Client
+
+	// application id "constant" memory-only cache
+	appTypeId      *string
+	appTypeIdMutex sync.Mutex
+)
+
+const Forever time.Duration = 24 * time.Hour * 30
+
+type Cacheable interface {
+	models.Account | clients.AccountDetailsAWS | clients.AzureTenantId
+}
+
+func Initialize() {
+	if config.Application.Cache.Type == "redis" {
+		redisEnabled = true
+		log.Logger.Info().Bool("cache", true).Msg("Initializing redis application cache")
+
+		// register all Cacheable types
+		gob.Register(&models.Account{})
+		gob.Register(&clients.AccountDetailsAWS{})
+
+		client = redis.NewClient(&redis.Options{
+			Addr:     config.RedisHostAndPort(),
+			Username: config.Application.Cache.Redis.User,
+			Password: config.Application.Cache.Redis.Password,
+			DB:       config.Application.Cache.Redis.DB,
+		})
+	} else {
+		log.Logger.Info().Bool("cache", true).Msg("No application cache in use")
+	}
+}
+
+func FindAppTypeId(_ context.Context) (string, error) {
+	appTypeIdMutex.Lock()
+	defer appTypeIdMutex.Unlock()
+
+	if appTypeId == nil {
+		return "", ErrNotFound
+	}
+	return *appTypeId, nil
+}
+
+func SetAppTypeId(_ context.Context, value string) error {
+	appTypeIdMutex.Lock()
+	defer appTypeIdMutex.Unlock()
+
+	appTypeId = &value
+	return nil
+}
+
+func Prefix[T Cacheable](value *T) (string, error) {
```

Comment:
Looks weird to use generics and then still switch by type.
Why is it better then having SetXXX function? :thinking:

Reply:
Yeah I rebased the code to use interface normally instead of constraint. This is indeed more natural and better, good point.